api/v1alpha1/clusterextension_types.go

```diff
@@ -18,53 +18,35 @@ package v1alpha1
-	// disastrous results such as data loss.
-	UpgradeConstraintPolicyIgnore UpgradeConstraintPolicy = "Ignore"
-)
+// +kubebuilder:validation:XValidation:rule="self.sourceType=='package' && has(self.__package__)",message="sourceType must match populated union field"
+//
+// ClusterExtensionSource defines the source for this ClusterExtension, right now, only a package is supported.
+type ClusterExtensionSource struct {
+	//+kubebuilder:validation:Enum:=package
+	//+kubebuilder:validation:Required
+	// SourceType is the discriminator for the source type
+	SourceType string `json:"sourceType"`
+
+	// Package defines a reference for a bundle in a catalog defined by a name and a version and/or channel
+	Package *SourcePackage `json:"package,omitempty"`
+}
 
 // ClusterExtensionSpec defines the desired state of ClusterExtension
 type ClusterExtensionSpec struct {
-	//+kubebuilder:validation:MaxLength:=48
-	//+kubebuilder:validation:Pattern:=^[a-z0-9]+(-[a-z0-9]+)*$
-	PackageName string `json:"packageName"`
-
-	//+kubebuilder:validation:MaxLength:=64
-	//+kubebuilder:validation:Pattern=`^(\s*(=||!=|>|<|>=|=>|<=|=<|~|~>|\^)\s*(v?(0|[1-9]\d*|[x|X|\*])(\.(0|[1-9]\d*|x|X|\*]))?(\.(0|[1-9]\d*|x|X|\*))?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?)\s*)((?:\s+|,\s*|\s*\|\|\s*)(=||!=|>|<|>=|=>|<=|=<|~|~>|\^)\s*(v?(0|[1-9]\d*|x|X|\*])(\.(0|[1-9]\d*|x|X|\*))?(\.(0|[1-9]\d*|x|X|\*]))?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?)\s*)*$`
 	//+kubebuilder:Optional
-	// Version is an optional semver constraint on the package version. If not specified, the latest version available of the package will be installed.
-	// If specified, the specific version of the package will be installed so long as it is available in any of the content sources available.
-	// Examples: 1.2.3, 1.0.0-alpha, 1.0.0-rc.1
 	//
-	// For more information on semver, please see https://semver.org/
-	Version string `json:"version,omitempty"`
+	// paused controls the management state of the ClusterExtension. If paused, it will be ignored by the ClusterExtension controller.
+	Paused bool `json:"paused,omitempty"`
 
-	//+kubebuilder:validation:MaxLength:=48
-	//+kubebuilder:validation:Pattern:=^[a-z0-9]+([\.-][a-z0-9]+)*$
-	// Channel constraint definition
-	Channel string `json:"channel,omitempty"`
+	// source of ClusterExtension to be installed
+	Source ClusterExtensionSource `json:"source"`
 
-	//+kubebuilder:validation:Enum:=Enforce;Ignore
-	//+kubebuilder:default:=Enforce
 	//+kubebuilder:Optional
 	//
-	// Defines the policy for how to handle upgrade constraints
-	UpgradeConstraintPolicy UpgradeConstraintPolicy `json:"upgradeConstraintPolicy,omitempty"`
+	// skipCRDUpgradeSafetyCheck specifies whether the CRD upgrade safety checks should be skipped when attempting to install the ClusterExtension
+	SkipCRDUpgradeSafetyCheck bool `json:"skipCRDUpgradeSafetyCheck,omitempty"`
```

Comment:
I think @rashmigottipati is working on some changes that are going to be more extensible for future preflight checks. I think it would be okay to keep this field out of these changes

Reply:
Oh, I see - this PR still has the extension type. I forgot https://github.com/operator-framework/operator-controller/pull/820 has not merged yet :)

Reply:
Ignore the previous comment. I accidentally commented this on the wrong review thread.